lib/pos_mempool.go

```diff
@@ -0,0 +1,391 @@
+	if err := dmp.addTransactionNoLock(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	if err := dmp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) addTransactionNoLock(txn *MempoolTx) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+	txnFee := txn.Tx.TxnFeeNanos
+	latestBlockHeight := dmp.latestBlockNode.Height
+
+	// Validate that the user has enough balance to cover the transaction fees.
+	spendableBalanceNanos, err := dmp.latestBlockView.GetSpendableDeSoBalanceNanosForPublicKey(userPk.ToBytes(), latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "CheckBalanceIncrease: Problem getting spendable balance")
+	}
+	if err := dmp.ledger.CanIncreaseBalance(*userPk, txnFee, spendableBalanceNanos); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem checking balance increase for transaction with"+
+			"hash %v, fee %v", txn.Tx.Hash(), txnFee)
+	}
+
+	// If we get here, it means that the transaction's sender has enough balance to cover transaction fees. We can now
+	// add the transaction to mempool.
+	if err := dmp.txnRegister.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding txn to register")
+	}
+
+	// We update the reserved balance to include the newly added transaction's fee.
+	dmp.ledger.IncreaseBalance(*userPk, txnFee)
+
+	// Emit an event for the newly added transaction.
+	if dmp.emitEvents {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventAdd,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// RemoveTransaction is the main function for removing a transaction from the mempool.
+func (dmp *PosMempool) RemoveTransaction(txn *MempoolTx) error {
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.RemoveTransaction: ")
+	}
+
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	return dmp.removeTransactionNoLock(txn)
+}
+
+func (dmp *PosMempool) removeTransactionNoLock(txn *MempoolTx) error {
+	// First, sanity check our reserved balance.
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+
+	// Remove the transaction from the register.
+	if err := dmp.txnRegister.RemoveTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.RemoveTransaction: Problem removing txn from register")
+	}
+	// Decrease the appropriate ledger's balance by the transaction fee.
+	dmp.ledger.DecreaseBalance(*userPk, txn.Fee)
+
+	// Emit an event for the removed transaction.
+	if dmp.emitEvents {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventRemove,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// GetTransaction returns the transaction with the given hash if it exists in the mempool. This function is thread-safe.
+func (dmp *PosMempool) GetTransaction(txHash *BlockHash) *MempoolTx {
```

Comment:
Just curious. Would it make sense to return *MsgDeSoTxn here instead? In a way, the MempoolTx construct is an implementation detail of the mempool. If we can get it to a point where consensus doesn't know about the struct, then we get better separation of concerns, which will make it easier for us decouple/move the mempool code into its own package later on.
```suggestion
func (dmp *PosMempool) GetTransaction(txHash *BlockHash) *MsgDeSoTxn {
```